[red-knot] Detect semantic syntax errors

ntBre opened the pull request:
Summary
--

This PR extends semantic syntax error detection to red-knot. The main changes here are:

1. Adding `SemanticSyntaxChecker` and `Vec<SemanticSyntaxError>` fields to the `SemanticIndexBuilder`
2. Calling `SemanticSyntaxChecker::visit_stmt` and `visit_expr` in the `SemanticIndexBuilder`'s `visit_stmt` and `visit_expr` methods
3. Implementing `SemanticSyntaxContext` for `SemanticIndexBuilder`
4. Adding new mdtests to test the context implementation and show diagnostics

(3) is definitely the trickiest and required (I think) a minor addition to the `SemanticIndexBuilder`. I tried to look around for existing code performing the necessary checks, but I definitely could have missed something or misused the existing code even when I found it.

There's still one TODO around `global` statement handling. I don't think there's an existing way to look this up, but I'm happy to work on that here or in a separate PR. This currently only affects detection of one error (`LoadBeforeGlobalDeclaration` or [PLE0118](https://docs.astral.sh/ruff/rules/load-before-global-declaration/) in ruff), so it's not too big of a problem even if we leave the TODO.

Test Plan
--

New mdtests, as well as new errors for existing mdtests

ntBre commented on crates/red_knot_python_semantic/resources/mdtest/comprehensions/basic.md:
```diff
@@ -128,6 +128,8 @@ class AsyncIterable:
         return AsyncIterator()
 
 # revealed: @Todo(async iterables/iterators)
+# error: [invalid-syntax] "cannot use an asynchronous comprehension outside of an asynchronous function on Python 3.9 (syntax was added in 3.11)"
+# error: [invalid-syntax] "`asynchronous comprehension` outside of an asynchronous function"
```
That PR does not fix the backticks around "asynchronous comprehension." I can just fix that here since red-knot will be the only place this message is actually used.

ntBre commented:
The `mypy_primer` results look like true positives to me, although it's a bit unfortunate because it seems clear that this file is meant to be split up somehow: 

https://github.com/laowantong/paroxython/blob/44e14b17965c6b02871db4c9ed6acc8716c2740d/examples/idioms/programs_with_labels.py#L690-L695

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/annotations/invalid.md:
```diff
@@ -68,8 +68,12 @@ def _(
     i: not 1,  # error: [invalid-type-form] "Unary operations are not allowed in type expressions"
     j: lambda: 1,  # error: [invalid-type-form] "`lambda` expressions are not allowed in type expressions"
     k: 1 if True else 2,  # error: [invalid-type-form] "`if` expressions are not allowed in type expressions"
+    # error: [invalid-syntax] "`await` statement outside of a function"
+    # error: [invalid-syntax] "`await` outside of an asynchronous function"
     l: await 1,  # error: [invalid-type-form] "`await` expressions are not allowed in type expressions"
+    # error: [invalid-syntax] "`yield` statement outside of a function"
     m: (yield 1),  # error: [invalid-type-form] "`yield` expressions are not allowed in type expressions"
+    # error: [invalid-syntax] "`yield from` statement outside of a function"
```
maybe let's just avoid the syntax errors by putting the entire `_` function inside an inner function? I think the syntax-error diagnostics here distract a bit from what this is meant to be testing

ntBre replied:
Sounds good! I was on the fence between marking the errors or editing the test cases and went for leaving the test code untouched, but this definitely makes sense, especially for these easy-to-fix cases.

What did you think about the `match` cases in `star.md`? I actually started there and was going to edit them so that the name capture was last, but I wasn't sure if the `or` pattern was also important.

```diff
diff --git a/a.py b/b.py
index f39fe59..2dd1a04 100644
--- a/a.py
+++ b/b.py
@@ -5,7 +5,7 @@ match 42:
         ...
     case [O]:
         ...
-    case P | Q:  # error: [invalid-syntax] "name capture `P` makes remaining patterns unreachable"
-        ...
     case object(foo=R):
         ...
+    case P:
+        ...
```

AlexWaygood replied:
Ooh, thanks, I didn't spot that! Yes, it is important that that's an `Or` pattern. The LHS and RHS of the `|` don't necessarily need to be `Name`s though -- you could add two classes to the mdtest:

```py
class Foo: ...
class Bar: ...
```

and then change the pattern to `Foo() | Bar()`?

ntBre replied:
Actually in the `P` and `Q` case, it does look important for them to be `Name`s? They're used down below in in a `possibly-unresolved-reference` error:

https://github.com/astral-sh/ruff/blob/454ad15aee13e5ac2fc9b67ff5907658764207df/crates/red_knot_python_semantic/resources/mdtest/import/star.md?plain=1#L244-L246

The shortest diff preserving that later assertion would probably be nesting them inside something else like `case object(p=P) | object(q=Q)`, but I'm happy to do whatever best preserves your original intentions with the tests.

Actually that suggestion causes yet another syntax error in CPython that I didn't cover:

```pycon
>>> match 42:
...     case object(p=P) | object(q=Q): ...
...
  File "<python-input-0>", line 2
    case object(p=P) | object(q=Q): ...
         ^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: alternative patterns bind different names
```

and also causes an `unresolved-reference` to R instead of a `possibly-unresolved-reference`.

AlexWaygood replied:
Ah, feel free to leave `star.md` as you have it then, don't worry too much about it! Would be great to open a followup issue about this and https://github.com/astral-sh/ruff/pull/17463#discussion_r2050872442 so we don't forget about it, though. I can look at it next week.

ntBre replied:
Opened #17522 and #17523 and added the new syntax error to #17412!

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comprehensions/basic.md:
```diff
@@ -128,6 +128,8 @@ class AsyncIterable:
 # revealed: @Todo(async iterables/iterators)
+# error: [invalid-syntax] "cannot use an asynchronous comprehension outside of an asynchronous function on Python 3.9 (syntax was added in 3.11)"
+# error: [invalid-syntax] "`asynchronous comprehension` outside of an asynchronous function"
 [reveal_type(x) async for x in AsyncIterable()]
```
similarly here, I don't think the intent here was for this to test our resilience to invalid syntax (the test case is titled "Basic" 😄), so I'd be inclined to do this to avoid the syntax error:

```suggestion
async def _():
    [reveal_type(x) async for x in AsyncIterable()]
```

although it probably _is_ a good idea for us to add a _separate_ test case that demonstrates that we still infer types correctly even if there is invalid syntax!

ntBre replied:
Updated! Leaving this unresolved for now in case we also want to add the separate test case. I'm guessing that could go in `invalid_syntax.md` next to this file? But maybe somewhere else is better.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comprehensions/basic.md:
```diff
@@ -145,6 +147,8 @@ class Iterable:
     def __iter__(self) -> Iterator:
         return Iterator()
 
+# error: [invalid-syntax] "cannot use an asynchronous comprehension outside of an asynchronous function on Python 3.9 (syntax was added in 3.11)"
+# error: [invalid-syntax] "`asynchronous comprehension` outside of an asynchronous function"
 # revealed: @Todo(async iterables/iterators)
 [reveal_type(x) async for x in Iterable()]
```
here too

```suggestion
async def _():
    # revealed: @Todo(async iterables/iterators)
    [reveal_type(x) async for x in Iterable()]
```

AlexWaygood reviewed (commented):
🥳

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2255,104 @@ where
+    }
+
+    fn in_module_scope(&self) -> bool {
+        self.scopes[self.current_scope()].kind().is_module()
```
we already have a method for this one ;)

```suggestion
        self.current_scope_is_global_scope()
```

but maybe we should just delete that method, move the implementation of that method to the trait method here, and refactor the callsites of the method to use the trait method?

```suggestion
        self.scope_stack.len() == 1
```

ntBre replied:
I figured I would miss at least one existing method, thanks!

Hmm, I was worried about having to import the trait everywhere this method is used, but so far it's only used in this file. I'll move the implementation to the trait then!

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2255,104 @@ where
+    }
+
+    fn in_notebook(&self) -> bool {
+        source_text(self.db.upcast(), self.file).is_notebook()
```
```suggestion
        self.file.source_type(self.db.upcast()).is_ipynb()
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/diagnostics/semantic_syntax_errors.md:
````diff
@@ -0,0 +1,143 @@
+```py
+from __future__ import annotations
+
+# error: [invalid-type-form] "Named expressions are not allowed in type expressions"
+# error: [invalid-syntax] "named expression cannot be used within a type annotation"
````
hmm, yeah, ideally we would get rid of the `invalid-type-form` error here; it's just redundant noise to the user to have two diagnostics for the same problem. I think @MichaReiser has mentioned in the past that he wanted a global solution to make sure we didn't emit any red-knot diagnostics from type inference on nodes that were already known to be invalid syntax. Would be another thing that would be good to file a followup issue for IMO.

ntBre replied:
Opened #17524!

MichaReiser replied:
Yes, there's a TODO in the `InferContext` to suppress diagnostics in that case. I'm not entirely sure yet how that should work but isn't something that should block this PR.

AlexWaygood reviewed (approved):
This is great, thank you!

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/diagnostics/semantic_syntax_errors.md:
Does this cover all syntax errors that we detect or does it only cover a subset of it? I think it'd be useful to have at least one test case to cover each syntax error (could be done in a follow-up) so that any changes to the red-knot's semantic model is tested against these checks.

ntBre replied:
I believe this subset should cover every context method, which should be the main interaction with the semantic model. But it definitely wouldn't hurt to cover every error variant too!

dhruvmanila replied:
Makes sense. This is not urgent and doesn't need to block this PR but something good to have.

ntBre replied:
I might create a help-wanted issue, if that's not too lazy of me :laughing: Only a subset of errors are covered by ruff integration tests too. The rest rely on inline tests.

dhruvmanila commented on crates/red_knot_python_semantic/resources/mdtest/import/star.md:
Is it possible to avoid raising the syntax errors in this file by e.g., separating out the match statements to avoid unreachable patterns? cc @AlexWaygood

AlexWaygood replied:
yes, should definitely be possible -- I'll do it as a followup (https://github.com/astral-sh/ruff/issues/17523)

dhruvmanila commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -83,6 +89,7 @@ pub(super) struct SemanticIndexBuilder<'db> {
 
     /// Flags about the file's global scope
     has_future_annotations: bool,
+    seen_module_docstring_boundary: bool,
```
At some point in the future, it would be useful to combine these flags into a `bitflags` or similar to avoid multiple fields.

dhruvmanila commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2251,104 @@ where
 }
 
+impl SemanticSyntaxContext for SemanticIndexBuilder<'_> {
+    fn seen_docstring_boundary(&self) -> bool {
+        self.seen_module_docstring_boundary
+    }
```
nit: should this method be named `seen_module_docstring_boundary` to explicitly state that it's for the module docstring and not other (function, class, etc.) docstring?

ntBre replied:
Ah, that is what the ruff `SemanticModel`'s method is called too. I'll update it!

dhruvmanila reviewed (approved):
Looks good!

ntBre commented:
Thanks for the reviews! I think I've addressed all of the feedback we wanted to handle here and opened issues for the other follow-ups. I'll leave this open a bit longer in case anyone else wants to have a look, especially Micha since he's been pinged a couple of times here.

MichaReiser commented on crates/ruff_db/src/diagnostic/mod.rs:
```diff
@@ -845,3 +845,16 @@ pub fn create_unsupported_syntax_diagnostic(
+    file: File,
+    err: &ruff_python_parser::semantic_errors::SemanticSyntaxError,
+) -> Diagnostic {
+    let mut diag = Diagnostic::new(DiagnosticId::InvalidSyntax, Severity::Error, "");
+    let span = Span::from(file).with_range(err.range);
+    diag.annotate(Annotation::primary(span).message(err.to_string()));
+    diag
```
I don't think it's good practice to create diagnostics without a main message. I suggest passing `err.to_string()` to the `Diagnostic::new` (instead of "") and instead use no message in the annotation (the annotation's message should describe the annotated part). 

We should make the same change for unsupported syntax and parse diagnostics but we can do this as a seprate PR (unless @BurntSushi thinks what we have now is better)

MichaReiser commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -1593,8 +1593,12 @@ pub trait SemanticSyntaxContext {
 
-    /// Returns the source text under analysis.
-    fn source(&self) -> &str;
+    /// Call `f` with the source text under analysis and return its result.
+    ///
+    /// This somewhat awkward API allows for cases where retrieving the source text requires an
+    /// intermediate reference that can't be returned directly (such as red-knot's `source_text`
+    /// function).
+    fn with_source<T>(&self, f: impl FnOnce(&str) -> T) -> T;
```
I don't fully remember the motivation for having `SourceText` and `SourceTextInner` instead of annotating the query with `return_ref` and replacing `SourceText` with `SourceTextInner`. 

We could try changing the definition as a separate PR.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -87,6 +88,10 @@ pub fn check_types(db: &dyn Db, file: File) -> TypeCheckDiagnostics {
         diagnostics.extend(result.diagnostics());
     }
 
+    for error in index.semantic_syntax_errors() {
```
Adding the diagnostic sementically fits better into `check_file_impl` because the diagnsotics aren't related to type checking. On the other hand, it doesn't really make a difference. That's why I think either way is fine (up to you)

ntBre replied:
Ah, is the `semantic_index` call cached with salsa? I thought I had to do it here because this is where the 
 `SemanticIndex` holding the errors was available, but if I can just call `semantic_index` again in `check_file_impl`, I can definitely move it.

ntBre replied:
Oh, `check_file_impl` is actually in a different crate, and `semantic_index` is `pub(crate)`. I think I'll just leave it here.

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1025,15 +1035,26 @@ impl<'db> SemanticIndexBuilder<'db> {
     fn visit_stmt(&mut self, stmt: &'ast ast::Stmt) {
+        self.with_semantic_checker(|semantic, context| semantic.visit_stmt(stmt, context));
+
+        self.seen_module_docstring_boundary = true;
```
This seems so simple that, arguably, we might just want to move this logic to the `SemanticSyntaxChecker` (the integration complexity is higher)

ntBre replied:
I moved this field onto the `SemanticSyntaxChecker`. I think the previous problem was from also trying to delete the flag from ruff's `SemanticModel`. So there's a slight duplication with ruff tracking it separately, but it's just a single `bool` and allows deleting a whole context method.

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2251,102 @@ where
+    }
+
+    fn in_module_scope(&self) -> bool {
+        self.scope_stack.len() == 1
```
You can call https://github.com/astral-sh/ruff/blob/be54b840e98a35069ab53746996684c887cc102c/crates/red_knot_python_semantic/src/semantic_index/builder.rs#L159-L161

AlexWaygood replied:
Brent is deleting that method as part of this PR; current callsites of that method are being switched over to use the `in_module_scope` method being added here

ntBre replied:
Alex actually suggested deleting that method in favor of this one and inlining its implementation here. Happy to revert if you prefer the old one, though.

ntBre replied:
Wow I should have refreshed before commenting! Thanks @AlexWaygood!

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/snapshots/semantic_syntax_errors.md_-_Semantic_syntax_error_diagnostics_-_`async`_comprehensions_in_synchronous_comprehensions_-_Python_3.10.snap:
```diff
@@ -0,0 +1,46 @@
+4 | async def f():
+5 |     # error: 19 [invalid-syntax] "cannot use an asynchronous comprehension outside of an asynchronous function on Python 3.10 (syntax...
+6 |     return {n: [x async for x in elements(n)] for n in range(3)}
+  |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use an asynchronous comprehension outside of an asynchronous function on Python 3.10 (syntax was added in 3.11)
```
I _think_ the "synchronous function" that the async list comprehension is immediately nested inside of here is the function-like scope created by the synchronous dict comprehension. But I agree that it would be great if we could have a clearer error and a multi-span diagnostic

ntBre replied:
Yes, `f` is async, but the version-related part is that you can't use an async comprehension inside of a sync comprehension before 3.11. This is the error message I'm trying to clean up in https://github.com/astral-sh/ruff/pull/17460, so I definitely agree that it's confusing!

That's a good point about the multi-span diagnostic too. We could definitely highlight the outer sync comprehension, as well as the inner async one.

AlexWaygood commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2251,102 @@ where
+    }
+
+    fn in_notebook(&self) -> bool {
+        self.file.source_type(self.db.upcast()).is_ipynb()
```
oops, I think that's what @ntBre had originally but I suggested doing it this way. Sorry @ntBre! (Micha almost certainly knows best here)

ntBre replied:
No worries, I think `source_text` makes more sense in combination with the suggestion to cache it down below than it did when I wrote it!

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -87,6 +88,10 @@ pub fn check_types(db: &dyn Db, file: File) -> TypeCheckDiagnostics {
         diagnostics.extend(result.diagnostics());
     }
 
+    for error in index.semantic_syntax_errors() {
+        diagnostics.push(create_semantic_syntax_diagnostic(file, error));
+    }
```
Oh, I assumed it's a vec. 

Would it work if you implement the `Extend` trait?

ntBre replied:
I went with the `extend_diagnostics` method approach. The implementation is the same as `impl Extend<Diagnostic>`, but I had to call it in the qualified `Extend::extend` form to differentiate from the current `extend` method, so I thought it looked nicer this way.

ntBre commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2251,102 @@ where
+        Program::get(self.db).python_version(self.db)
+    }
+
+    fn with_source<T>(&self, f: impl FnOnce(&str) -> T) -> T {
+        let src = source_text(self.db.upcast(), self.file);
+        f(src.as_str())
+    }
+
```
I haven't profiled to be sure, but they don't look like they're called particularly often. `is_notebook` is usually gated by `in_module_scope`, but it is used in the `async` checks in global scope. `with_source` is only used for extracting a duplicate `match` mapping key, so it's almost never called.

But I still like the idea of the `LazyCell` and restoring the simple signature for the `source` method, so I think I'll do that.

ntBre replied:
I'm actually having some trouble with the `LazyCell`. With this straightforward approach:

```rust
        let mut builder = Self {
            db,
            file,
            source_text: LazyCell::new(|| source_text(db, file)),
            // ...
        };
```

the compiler is complaining that it found a closure capturing variables instead of an `fn` item. Am I doing something silly here, or should I try something else?

Obviously `LazyCell` closures can generally capture variables because there's an example in ruff, but maybe it's because I'm trying to return it.

MichaReiser replied:
I think the probleme here is more that you can't name the closure type when declaring `SemanticIndexBuilder`. You can try to use `OnceCell` instead

ntBre commented:
I'm guessing (hoping) the horrible performance degradation was from my naive use of `is_file_open`. I started caching that on the builder and also merged main in case it was something else.

ntBre commented:
One more guess and then I'll set up for local benchmarks :sweat_smile: I'm not getting much from the salsa flamegraphs on codspeed.

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2268,6 +2289,100 @@ where
+    }
+
+    fn report_semantic_error(&self, error: SemanticSyntaxError) {
+        if self.db.is_file_open(self.file) {
```
You could consider changing the `semantic_checker` field on the `SemanticIndexBuilder` to be an `Option<SemanticSyntaxChecker>` and only initialize it with `Some` if the file is open. `with_semantic_checker` is a no-op if the checker is `None`.

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -2228,6 +2251,102 @@ where
+    }
+
+    fn report_semantic_error(&self, error: SemanticSyntaxError) {
+        self.semantic_syntax_errors.borrow_mut().push(error);
```
You want to cache it. It's at least as expensive as a hash set lookup

MichaReiser replied:
The regression went away, no? I no longer see the 33% regression. A 1% regresssion doesn't sound that unreasonable to me

ntBre replied:
It went away after I reverted my first two attempts and moved the check into `report_semantic_error`. The benchmark just finished on the `Option` case and it seems to have brought back the regression. Do you have an intuition why calling `is_file_open` in `SemanticIndexBuilder::new` is such a disaster?

MichaReiser replied:
Not really. `is_file_open` is a "cheap" hash set lookup in the benchmark case 

https://github.com/astral-sh/ruff/blob/9c47b6dbb0f29369fdf38d5048276994b2e09e49/crates/red_knot_project/src/lib.rs#L316-L324

It also doesn't introduce any new query dependencies that would be expensive to validate.

I can try to debug this tomorrow.

ntBre replied:
Thanks, and sorry for the trouble. I'll push my other changes, I think everything else should be resolved.

MichaReiser replied:
Okay. I think I know what's happening but I don't have a solution for this yet. 

Salsa assigns values to different durabilities. This allows it to short-circuit some expensive checks if nothing of that durability has changed. In our case, we use a *higher* durability for files from the standard library because they can never change (all files in typeshed are shipped as part of the red knot binary). 

That means, the `semantic_index` query for files from typeshed currently have a high durability. However, this is no longer the case if we call `is_file_open` because it reads `open_files` which has a low durability -> Making the entire query having a low durability. 

The easiest fix is to change `is_file_open` to always return `false` if `file.path(db).is_vendored()`. However, this still has the downside that it reduces the durability of `semantic_index` for all third-party files to `low`. 

I've to think a bit more about what the proper fix is here. Maybe going back to when you emit the diagnostic isn't that bad after all.

MichaReiser replied:
I went ahead and created an issue for this "downgrading" problem as this might is also a problem for type inference diagnostics. To unblock this PR, I suggest to go back to your implementation where you lazily check if the file is open before reporting the diagnostic.

ntBre replied:
Thanks for tracking down the issue! I've reverted to having the check in `report_semantic_error`.

I'll make sure the benchmarks pass this time and then get ready to merge, if that's good with you!

MichaReiser commented on crates/red_knot_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -98,6 +104,10 @@ pub(super) struct SemanticIndexBuilder<'db> {
     expressions_by_node: FxHashMap<ExpressionNodeKey, Expression<'db>>,
     imported_modules: FxHashSet<ModuleName>,
     eager_bindings: FxHashMap<EagerBindingsKey, ScopedEagerBindingsId>,
+    python_version: PythonVersion,
+    semantic_checker: SemanticSyntaxChecker,
```
We should move those fields into their own group (or to the `builder state` group at the top). The fields at the bottom are reserved for fields that match the fields on `SemanticIndex`
